src/pybind/mgr/volumes/module.py

```diff
@@ -205,26 +205,6 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
             'desc': "Set MDS pinning policy for subvolume",
             'perm': 'rw'
         },
-        {
-            'cmd': 'fs subvolume snapshot protect '
-                   'name=vol_name,type=CephString '
-                   'name=sub_name,type=CephString '
-                   'name=snap_name,type=CephString '
-                   'name=group_name,type=CephString,req=false ',
-            'desc': "Protect snapshot of a CephFS subvolume in a volume, "
-                    "and optionally, in a specific subvolume group",
-            'perm': 'rw'
-        },
-        {
-            'cmd': 'fs subvolume snapshot unprotect '
-                   'name=vol_name,type=CephString '
-                   'name=sub_name,type=CephString '
-                   'name=snap_name,type=CephString '
-                   'name=group_name,type=CephString,req=false ',
-            'desc': "Unprotect a snapshot of a CephFS subvolume in a volume, "
-                    "and optionally, in a specific subvolume group",
-            'perm': 'rw'
```

Comment:
I'm afraid the horse has left the barn on this one. ceph-csi is already using this interface? We should still accept these operations but do nothing I suppose.

Reply:
@vshankar and me had this discussion,
- ceph-csi is yet to add code to support cephfs snapshots, which includes these calls
- But, between ceph versions there would still be a need to protect/unprotect, and ceph-csi, when it adds code to support the same, may need to handle the required calls

We possibly should add a feature flag to `subvolume info` that helps detect if protect/unprotect is required, to keep things clean. Absence of the feature flag indicates it requires protect/unprotect. Thoughts?

Reply:
> @vshankar and me had this discussion,
> 
> * ceph-csi is yet to add code to support cephfs snapshots, which includes these calls

right.

> * But, between ceph versions there would still be a need to protect/unprotect, and ceph-csi, when it adds code to support the same, may need to handle the required calls

sorry, it's not 100% clear to me why [un]protect is needed b/w ceph versions since ceph-csi is not _yet_ there.

> 
> We possibly should add a feature flag to `subvolume info` that helps detect if protect/unprotect is required, to keep things clean. Absence of the feature flag indicates it requires protect/unprotect. Thoughts?

Reply:
> > @vshankar and me had this discussion,
> > 
> > * ceph-csi is yet to add code to support cephfs snapshots, which includes these calls
> 
> right.
> 
> > * But, between ceph versions there would still be a need to protect/unprotect, and ceph-csi, when it adds code to support the same, may need to handle the required calls
> 
> sorry, it's not 100% clear to me why [un]protect is needed b/w ceph versions since ceph-csi is not _yet_ there.

The case I can think of is a cluster running say v15.2.[0-3] and using ceph-CSI in the future. In these clusters if [un]protect calls are not made, the clone operation will fail.

The alternative to this is to let them fail, and call for a version support matrix by ceph-csi that is has this feature removed in it.

> 
> > We possibly should add a feature flag to `subvolume info` that helps detect if protect/unprotect is required, to keep things clean. Absence of the feature flag indicates it requires protect/unprotect. Thoughts?

Reply:
> The case I can think of is a cluster running say v15.2.[0-3] and using ceph-CSI in the future. In these clusters if [un]protect calls are not made, the clone operation will fail.

Ummm... unless this is backported and CSI uses the latest nautilus.

Reply:
> > The case I can think of is a cluster running say v15.2.[0-3] and using ceph-CSI in the future. In these clusters if [un]protect calls are not made, the clone operation will fail.
> 
> Ummm... unless this is backported and CSI uses the latest nautilus.

I maybe missing something but,
- CSI could be used against the older version ceph clusters. Their support matrix includes nautilus and octopus (from a clone support POV).
- Backporting will ensure, say in the given example, v15.4 would no longer require [un]protect calls.
  - Is there a way to backport this to v15.[0-3].z? and, that is what you mean by backport? and,
  - if this is done users are expected to run the latest 'z' release for the feature (which I believe is acceptable)?

Reply:
> I maybe missing something but,
> 
> * CSI could be used against the older version ceph clusters. Their support matrix includes nautilus and octopus (from a clone support POV).
> * Backporting will ensure, say in the given example, v15.4 would no longer require [un]protect calls.
>   
>   * Is there a way to backport this to v15.[0-3].z? and, that is what you mean by backport? and,

Obviously not.

>   * if this is done users are expected to run the latest 'z' release for the feature (which I believe is acceptable)?

Right -- all this is assuming that by the time CSI has support for snap and clones, we'll already be in some z-release that CSI will rely on. My inclination towards dropping [un]protect is mainly to get rid of a good chunk of code from mgr/volumes since it's pretty much not required (it was added in the first place to mimic RBD calls, but then RBD moved away from that anyway).

Reply:
> > I maybe missing something but,
> > 
> > * CSI could be used against the older version ceph clusters. Their support matrix includes nautilus and octopus (from a clone support POV).
> > * Backporting will ensure, say in the given example, v15.4 would no longer require [un]protect calls.
> >   
> >   * Is there a way to backport this to v15.[0-3].z? and, that is what you mean by backport? and,
> 
> Obviously not.
> 
> > * if this is done users are expected to run the latest 'z' release for the feature (which I believe is acceptable)?
> 
> Right -- all this is assuming that by the time CSI has support for snap and clones, we'll already be in some z-release that CSI will rely on. My inclination towards dropping [un]protect is mainly to get rid of a good chunk of code from mgr/volumes since it's pretty much not required (it was added in the first place to mimic RBD calls, but then RBD moved away from that anyway).

So here are the options and why,
Why:
- CSI is currently the only consumer of the [un]protect API, and has not yet coded these up in their master branch
- By the time CSI makes a release with CephFS snapshots/clones support, we will have this PR back ported to respective 14.x and 15.y Ceph releases (this could be as soon as end of July'2020)
Options:
1. Let CSI call out minimum supported versions for CephFS clones as 14.x and 15.y
    - IOW, in versions where [un]protect is mandatory, CSI will error out during the clone operation as the snapshot is not protected
2. We make a contentious choice to provide a supported feature list (via `info` API say) for a subvolume, for CSI to take appropriate action on newer and older versions of nautilus and octopus w.r.t invoking [un]protect APIs

@batrick we (@vshankar and myself) started at (1) above. If (2) is better we should do it and leave the choice to CSI. Request your feedback regarding the same. Thanks.

Reply:
(1) sounds fine. Thanks for explaining.